File: targets/GATE/lib/dispatcher.py
import urllib.parse
import json, urllib.request, urllib.error
from openc3.interfaces.protocols.protocol import Protocol
from openc3.config.config_parser import ConfigParser
from openc3.utilities.logger import Logger

class Dispatcher(Protocol):
   """
   A WRITE protocol that inspects the command Packet (pre-encode),
   logs/sends a JSON summary to an external service, then returns
   the packet to continue the command pipeline.
   """

   def __init__(self, rest_endpoint:str, keycloak_identity=None, allow_empty_data=None):
      super().__init__(allow_empty_data)
      # Normalize endpoints like "localhost:8080/authorize" -> "http://localhost:8080/authorize"
      self.rest_endpoint = self._normalize_endpoint(rest_endpoint)
      self.keycloak_identity = keycloak_identity
      self.allow_empty_data = ConfigParser.handle_true_false_none(allow_empty_data)
      # Store raw response bytes from REST calls (set on every dispatch)
      self.last_response_bytes = None

   def _normalize_endpoint(self, rest_endpoint: str) -> str:
      """
      Ensures rest_endpoint is a proper URL:
      - Adds default scheme (http) if missing.
      - Supports scheme-relative inputs like //host:port/path.
      - Ensures path begins with '/' when a netloc exists.
      """
      if not rest_endpoint:
         raise ValueError("rest_endpoint must be a non-empty string")

      ep = rest_endpoint.strip()

      # If no explicit scheme delimiter is present, add http:// (or http: for scheme-relative)
      if '://' not in ep:
         ep = ('http:' + ep) if ep.startswith('//') else ('http://' + ep)

      parsed = urllib.parse.urlparse(ep)

      # If we have a host:port and a non-empty path, ensure it starts with '/'
      if parsed.netloc and parsed.path and not parsed.path.startswith('/'):
         parsed = parsed._replace(path='/' + parsed.path)

      normalized = urllib.parse.urlunparse(parsed)
      Logger.info(f"Dispatcher: Normalized REST endpoint to {normalized}")
      return normalized

   def write_packet(self, packet):
      """
      Called BEFORE encoding. Returning the Packet continues the write;
      returning self.STOP drops; returning self.DISCONNECT disconnects interface.
      """
      keycloak_id = self.keycloak_identity or "unknown"
      pkt_name = packet.packet_name
      tgt_name = packet.target_name
      stream_id_item = packet.get_item("CCSDS_STREAMID")
      func_code_item = packet.get_item("CCSDS_FC")
      stream_id = packet.read_item(stream_id_item)
      func_code = packet.read_item(func_code_item)
      
      # Let NOOP commands pass through without enforcing policy
      if func_code == 1:
         Logger.info(f"Dispatcher: Received NOOP command (FUNCTION_CODE=1); passing through without dispatch")
         return packet

      # Build a summary dictionary (match Aranya Gate's lib.rs CMDSummary)
      summary = {
         "keycloak_id": keycloak_id,
         "target": tgt_name,
         "packet_name": pkt_name,
         "stream_id": stream_id,
         "function_code": func_code
      }
      
      # Serialize to JSON
      summary_json = json.dumps(summary).encode('utf-8')
      Logger.info(f"Dispatcher: Prepared packet summary JSON: {summary_json}")
      Logger.info(f"Dispatcher: Sending packet summary to {self.rest_endpoint}")

      # Dispatch the summary and decide whether to continue
      should_continue = self.dispatch_packet(summary_json, packet)
      if not should_continue:
         Logger.warn("Dispatcher: Halting command pipeline as dispatch_packet returned False")
         return 'STOP'

      # Use REST bytes to populate SER_CMD before returing packet for encoding CMD (BLOCK 1024 bits = 128 bytes)
      try:
         resp = self.last_response_bytes or b''
         Logger.info(f"Dispatcher: Writing SER_CMD with {len(resp)} bytes from REST response")
         packet.write("SER_CMD", resp)
      except Exception as e:
         Logger.error(f"Dispatcher: Failed to set SER_CMD on packet: {e}")
         return 'STOP'

      # Optionally log what we received from the REST API
      if self.last_response_bytes is not None:
         Logger.info(f"Dispatcher: REST response bytes length={len(self.last_response_bytes)}")
         summary["serialized_command_bytes"] = list(self.last_response_bytes)
         Logger.info(f"Final packet: {summary}")

      Logger.info("Dispatcher: Packet summary dispatched and updated successfully")
      # IMPORTANT: return the SAME packet SCHEMA to continue the pipeline unchanged.
      # Updating existing packet items is allowed
      return packet

   def dispatch_packet(self, summary_json, packet):
      # Send to REST endpoint
      try:
         req = urllib.request.Request(
            self.rest_endpoint,
            data=summary_json,
            headers={'Content-Type': 'application/json'},
            method='POST'  # make POST explicit
         )
         with urllib.request.urlopen(req) as response:
            status = getattr(response, 'status', response.getcode())
            # Read raw bytes so we can accept application/octet-stream without corrupting data
            body_bytes = response.read()
            self.last_response_bytes = body_bytes
            # Try to get a content type for logging
            try:
               content_type = response.headers.get_content_type()
            except Exception:
               content_type = None

            if 200 <= status < 300:
               # For binary payloads, don't log content to avoid noise
               if content_type == 'application/octet-stream':
                  Logger.info(f"Dispatcher: Sent packet summary to {self.rest_endpoint} (HTTP {status}), received {len(body_bytes)} bytes (octet-stream)")
               else:
                  # Best-effort decode for text responses
                  preview = ''
                  try:
                     preview = body_bytes.decode('utf-8', errors='replace')
                  except Exception:
                     preview = f"<{len(body_bytes)} bytes>"
                  Logger.info(f"Dispatcher: Sent packet summary to {self.rest_endpoint} (HTTP {status}), response: {preview}")
               return True
            else:
               # Non-success: log with safe preview
               preview = ''
               try:
                  preview = body_bytes.decode('utf-8', errors='replace')
               except Exception:
                  preview = f"<{len(body_bytes)} bytes>"
               Logger.error(f"Dispatcher: Non-success status from {self.rest_endpoint}: HTTP {status}, response: {preview}")
               return False
      except urllib.error.HTTPError as e:
         # Read and store error body bytes safely
         try:
            body_bytes = e.read()
         except Exception:
            body_bytes = b''
         self.last_response_bytes = body_bytes
         try:
            body = body_bytes.decode('utf-8', errors='replace')
         except Exception:
            body = f"<{len(body_bytes)} bytes>"
         Logger.error(f"Dispatcher: HTTPError sending packet summary to {self.rest_endpoint}: HTTP {e.code}, response: {body}")
         return False
      except urllib.error.URLError as e:
         self.last_response_bytes = None
         Logger.error(f"Dispatcher: URLError sending packet summary to {self.rest_endpoint}, error: {e}")
         return False
      except Exception as e:
         self.last_response_bytes = None
         Logger.error(f"Dispatcher: Unexpected error sending packet summary to {self.rest_endpoint}: {e}")
         return False

File: targets/GATE/lib/dispatcher.rb
# targets/<YOUR_TARGET>/lib/dispatcher.rb
require 'uri'
require 'json'
require 'net/http'

module OpenC3
  # WRITE protocol: inspect/gate command packets and optionally populate SER_CMD
  class Dispatcher < Protocol
    # allow_empty_data follows COSMOS Protocol semantics (nil/true/false)
    def initialize(rest_endpoint, keycloak_identity = nil, allow_empty_data = nil)
      super(allow_empty_data)
      @rest_endpoint       = normalize_endpoint(rest_endpoint)
      @keycloak_identity   = keycloak_identity
      @last_response_bytes = nil
    end

    # ---- WRITE side ----
    # Called BEFORE encoding. Return the packet to continue, or :STOP / :DISCONNECT
    def write_packet(packet)
      keycloak_id = @keycloak_identity || 'unknown'
      pkt_name    = packet.packet_name
      tgt_name    = packet.target_name

      # Read CCSDS header items
      stream_id_item = packet.get_item('CCSDS_STREAMID')
      func_code_item = packet.get_item('CCSDS_FC')
      stream_id      = packet.read_item(stream_id_item)  # default :CONVERTED
      func_code      = packet.read_item(func_code_item)  # default :CONVERTED

      # Let NOOP pass through without remote check
      if func_code == 1
        Logger.info('Dispatcher: NOOP command; passing without dispatch')
        return packet
      end

      # Build JSON summary (must match your upstream service expectations)
      summary = {
        keycloak_id:   keycloak_id,
        target:        tgt_name,
        packet_name:   pkt_name,
        stream_id:     stream_id,
        function_code: func_code
      }
      summary_json = JSON.dump(summary)
      Logger.info("Dispatcher: Prepared JSON #{summary_json}")
      Logger.info("Dispatcher: POST -> #{@rest_endpoint}")

      # Call remote gate
      should_continue = dispatch_packet(summary_json)
      unless should_continue
        Logger.warn('Dispatcher: Gate denied or error; stopping pipeline')
        return :STOP  # COSMOS expects :STOP to halt write chain. :contentReference[oaicite:2]{index=2}
      end

      # Populate SER_CMD with raw response bytes (BLOCK field in command)
      begin
        resp = @last_response_bytes || ''.b
        Logger.info("Dispatcher: Writing SER_CMD with #{resp.bytesize} byte(s)")
        packet.write('SER_CMD', resp) # write into modeled field before encode
      rescue => e
        Logger.error("Dispatcher: Failed to set SER_CMD: #{e}")
        return :STOP
      end

      if @last_response_bytes
        Logger.info("Dispatcher: REST response bytes length=#{@last_response_bytes.bytesize}")
      end
      Logger.info("Dispatcher: OK for #{tgt_name} #{pkt_name}")
      packet
    end

    private

    # Normalize endpoint:
    # - add http:// if missing
    # - handle //host:port/path style
    # - ensure path begins with '/'
    def normalize_endpoint(raw)
      raise ArgumentError, 'rest_endpoint must be non-empty' if raw.nil? || raw.strip.empty?
      ep = raw.strip
      ep = ep.start_with?('//') ? "http:#{ep}" : "http://#{ep}" unless ep.include?('://')

      uri = URI.parse(ep)
      if uri.host && uri.path && !uri.path.empty? && !uri.path.start_with?('/')
        uri.path = "/#{uri.path}"
      end
      normalized = uri.to_s
      Logger.info("Dispatcher: Normalized REST endpoint to #{normalized}")
      normalized
    end

    # POST summary and cache raw response bytes for later use
    def dispatch_packet(summary_json)
      uri       = URI.parse(@rest_endpoint)
      http      = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == 'https')

      req = Net::HTTP::Post.new(uri.request_uri)
      req['Content-Type'] = 'application/json'
      req.body = summary_json

      begin
        res = http.request(req)
        body = res.body ? res.body.b : ''.b
        @last_response_bytes = body
        ctype = res['Content-Type']

        if res.is_a?(Net::HTTPSuccess)
          Logger.info("Dispatcher: Success (#{res.code}), content-type=#{ctype}")
          true
        else
          preview =
            if body && !body.empty?
              begin
                tmp = body.dup
                tmp.force_encoding('UTF-8')
                tmp.valid_encoding? ? tmp[0, 1024] : "<#{body.bytesize} bytes>"
              rescue
                "<#{body.bytesize} bytes>"
              end
            else
              '<empty>'
            end
          Logger.error("Dispatcher: HTTP #{res.code} from #{@rest_endpoint}: #{preview}")
          false
        end
      rescue Net::HTTPError => e
        @last_response_bytes = e.respond_to?(:response) && e.response ? (e.response.body || '').b : nil
        body_str =
          begin
            b = @last_response_bytes || ''.b
            s = b.dup; s.force_encoding('UTF-8')
            s.valid_encoding? ? s : "<#{b.bytesize} bytes>"
          rescue
            "<#{@last_response_bytes ? @last_response_bytes.bytesize : 0} bytes>"
          end
        Logger.error("Dispatcher: HTTPError to #{@rest_endpoint}: #{e} body=#{body_str}")
        false
      rescue => e
        @last_response_bytes = nil
        Logger.error("Dispatcher: Unexpected error to #{@rest_endpoint}: #{e}")
        false
      end
    end
  end
end
